java/arcs/core/util/UnionFind.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.util
+
+/**
+ * A union-find datastructure for computing equivalence classes
+ * consisting of `Element` instances.
+ */
+class UnionFind<Element> {
+    /**
+     * A node in the union find datastructure.
+     */
+    private inner class Node(e: Element, p: Node?) {
+        val element: Element = e
+        var parent: Node? = p
+    }
+
+    private val nodes = HashMap<Element, Node>()
+
+    /**
+     * Unifies the equivalence classes of elements [e1] and [e2]. If either
+     * element is not present in any set, it is added.
+     */
+    fun union(e1: Element, e2: Element) {
+        val e1Root = findRoot(e1)
+        val e2Root = findRoot(e2)
+        if (e1Root != e2Root) {
```

Comment:
Should this be looking for strict object-reference inequality? If so, consider `!==`

Reply:
Good point. Done.